editor/src/components/canvas/dom-lookup.ts

```diff
@@ -114,60 +93,66 @@ export function findFirstParentWithValidElementPathInner(
-        const validPathFromString = EP.fromString(validPath)
-        for (const staticAndDynamic of staticAndDynamicTargetElementPaths) {
-          if (EP.isDescendantOfOrEqualTo(staticAndDynamic.staticPath, validPathFromString)) {
-            const depthDiff =
-              EP.fullDepth(staticAndDynamic.staticPath) - EP.fullDepth(validPathFromString)
-            filteredValidPathsMappedToDynamic.push(
-              EP.nthParentPath(staticAndDynamic.dynamic, depthDiff),
-            )
-          }
-        }
-      }
-      break
-    case 'search-dom-only':
-      for (const validPath of validStaticElementPaths) {
-        const pathToAdd = staticAndDynamicTargetElementPaths.find(
-          (staticAndDynamic) => staticAndDynamic.static === validPath,
-        )?.dynamic
-        if (pathToAdd != null) {
-          filteredValidPathsMappedToDynamic.push(pathToAdd)
+// Take a DOM element, and try to find the nearest selectable path for it
+export function findFirstParentWithValidElementPath(
+  validDynamicElementPathsForLookup: Set<string> | 'no-filter',
+  target: Element,
+  parentSceneValidPathsCache: FindParentSceneValidPathsCache,
+): ElementPath | null {
+  const staticAndDynamicTargetElementPaths = getStaticAndDynamicElementPathsForDomElement(target)
+
+  const validStaticElementPaths = getValidStaticElementPathsForDomElement(
+    target,
+    parentSceneValidPathsCache,
+    validDynamicElementPathsForLookup,
+  )
+
+  let resultPath: ElementPath | null = null
+  let maxDepth = -1
+  for (const validPath of validStaticElementPaths) {
+    const validPathFromString = EP.fromString(validPath)
+    // We try to find a valid dynamic path with the deepest possible element path.
+    // We use two algorithms, and the one with the deeper result wins.
+
+    // 1. Go through all element paths from DOM and find the closest ancestor of the static paths which are also a valid path.
+    // When this ancestor is found, get the dynamic element path version of the static path, and step upwards
+    // the same number of steps in the hierarchy from there: this dynamic path can be the a valid target.
+    for (const staticAndDynamic of staticAndDynamicTargetElementPaths) {
+      if (EP.isDescendantOfOrEqualTo(staticAndDynamic.staticPath, validPathFromString)) {
+        const depthDiff =
+          EP.fullDepth(staticAndDynamic.staticPath) - EP.fullDepth(validPathFromString)
+        const pathToAdd = EP.nthParentPath(staticAndDynamic.dynamic, depthDiff)
+        if (maxDepth < EP.fullDepth(pathToAdd)) {
+          maxDepth = EP.fullDepth(pathToAdd)
+          resultPath = pathToAdd
         }
       }
-      break
-    default:
-      const _exhaustiveCheck: never = allowDescendantsFromPath
-  }
+    }
+
+    // 2. Start to traverse the DOM elements upwards in the hierarchy.
+    // When we find an element which is attached to a static path which is also in the valid path list,
+    // the dynamic version of that path is a valid target
```

Comment:
possibly minor but I think we can skip this walk up the tree if we know that the depth will not be able to be greater than maxDepth